pkg/hooks/manager.go

```diff
@@ -252,83 +257,169 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
 			return nil, err
 		}
 		domainSpecXML = result.GetDomainXML()
+	case hooksV1alpha3.Version:
+		client := hooksV1alpha3.NewCallbacksClient(conn)
+		result, err := client.OnDefineDomain(ctx, &hooksV1alpha3.OnDefineDomainParams{
+			DomainXML: domainSpecXML,
+			Vmi:       vmiJSON,
+		})
+		if err != nil {
+			log.Log.Reason(err).Error("Failed to call OnDefineDomain")
+			return nil, err
+		}
+		domainSpecXML = result.GetDomainXML()
 	default:
 		log.Log.Errorf("Unsupported callback version: %s", callback.Version)
 	}
 
 	return domainSpecXML, nil
 }
 
+func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cloudinit.CloudInitData) ([3][]byte, error) {
+	empty := [3][]byte{}
+	vmiJSON, err := json.Marshal(vmi)
+	if err != nil {
+		return empty, fmt.Errorf("failed to marshal VMI spec: %v, err: %v", vmi, err)
```

Comment:
nit: vmi is a big struct. Printing it may result in flooded log. I would just return the error message. The same might be considered when printing `cloudInitData` in the code below. But feel free to ignore if you disagree.

Reply:
That's how it was before, I just moved the code to make introducing the new v1alpha3 easier but I agree with you that that error is enough here. I'll fix and add a note in the commit log about the change.

Reply:
Not tested yet, just to clarify the change.

```diff
diff --git a/pkg/hooks/manager.go b/pkg/hooks/manager.go
index 36ced5aae..8ee154607 100644
--- a/pkg/hooks/manager.go
+++ b/pkg/hooks/manager.go
@@ -279,7 +279,7 @@ func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cl
        empty := [3][]byte{}
        vmiJSON, err := json.Marshal(vmi)
        if err != nil {
-               return empty, fmt.Errorf("failed to marshal VMI spec: %v, err: %v", vmi, err)
+               return empty, fmt.Errorf("failed to marshal VMI spec, err: %v", err)
        }

        // To be backward compatible to sidecar hooks still expecting to receive the cloudinit data as a
@@ -297,7 +297,7 @@ func preCloudInitIsoDataToJSON(vmi *v1.VirtualMachineInstance, cloudInitData *cl

        cloudInitDataJSON, err := json.Marshal(cloudInitData)
        if err != nil {
-               return empty, fmt.Errorf("failed to marshal CloudInitData: %v, err: %v", cloudInitData, err)
+               return empty, fmt.Errorf("failed to marshal CloudInitData, err: %v", err)
        }
```